### ISSUE TITLE
AutosklearnClassifier/SimpleClassificationPipeline uses default configuration of data preprocessing pipeline

### ISSUE THREAD

**Issue by rabsr** [automl/auto-sklearn#916]
## Describe the bug ##
Autosklearn always uses default configuration of data preprocessing pipeline for training or predict. I tried extending rescaling component. The custom scaler was always initialised but fit and transform of custom scaler was never called. AutosklearnClassifier/SimpleClassificationPipeline always uses default configuration of data preprocessor during fit and transform.

## To Reproduce ##
I extended rescaling component. Following is the extended component

```
from autosklearn.pipeline.components.base import AutoSklearnPreprocessingAlgorithm
from autosklearn.pipeline.components.data_preprocessing.rescaling.abstract_rescaling import Rescaling
from autosklearn.pipeline.constants import DENSE, SPARSE, UNSIGNED_DATA, INPUT
from scipy import sparse

class CustomScalerComponent(Rescaling, AutoSklearnPreprocessingAlgorithm):
    def __init__(self, random_state):
        print("Custom Scaler: Initialized")

        from sklearn.preprocessing import StandardScaler
        self.preprocessor = StandardScaler(copy=False)

    @staticmethod
    def get_properties(dataset_properties=None):
        return {'shortname': 'CustomScaler',
                'name': 'CustomScaler',
                'handles_regression': True,
                'handles_classification': True,
                'handles_multiclass': True,
                'handles_multilabel': True,
                'is_deterministic': True,
                # TODO find out if this is right!
                'input': (SPARSE, DENSE, UNSIGNED_DATA),
                'output': (INPUT,)}

    def fit(self, X, y=None):
        print("Custom Scaler: Fit")

        if sparse.isspmatrix(X):
            self.preprocessor.set_params(with_mean=False)

        return super(CustomScalerComponent, self).fit(X, y)

    def transform(self, X):
        print("Custom Scaler: Transform")
        return super(CustomScalerComponent, self).transform(X)
```

**Autosklearn examlple code**
```import autosklearn.classification
import sklearn
from autosklearn.pipeline.components.data_preprocessing.rescaling import add_rescaler
from mlsautoml.pipeline.components.data_preprocessing.rescaling.custom_scaler import CustomScalerComponent


def main():
    add_rescaler(CustomScalerComponent)
    # Generate data.
    from sklearn.datasets import load_breast_cancer
    from sklearn.model_selection import train_test_split

    X, y = load_breast_cancer(return_X_y=True)
    X_train, X_test, y_train, y_test = train_test_split(X, y)

    # Fit MLP classifier to the data.
    clf = autosklearn.classification.AutoSklearnClassifier(
        time_left_for_this_task=120,
        per_run_time_limit=20,
        include_estimators=['random_forest'],
        include_preprocessors=['no_preprocessing'],
        ensemble_size=1,
        tmp_folder="/tmp/custom_scaler",
        delete_tmp_folder_after_terminate=False,

    )
    print(clf)
    clf.fit(X_train, y_train, dataset_name='breast_cancer')

    # Print test accuracy and statistics.
    y_pred = clf.predict(X_test)
    print("accuracy: ", sklearn.metrics.accuracy_score(y_pred, y_test))
    print("\n\n+++++++++++++++++++++++++++++++++++++++++++Sprint+++++++++++++++++++++++++++++++++++++++++++")
    print(clf.sprint_statistics())
    print("\n\n+++++++++++++++++++++++++++++++++++++++++++Show Model+++++++++++++++++++++++++++++++++++++++++++")
    print(clf.show_models())
    
    import pandas as pd
    pd.DataFrame(clf.cv_results_).to_csv("custom_scaler.csv")


if __name__ == "__main__":
    main()
```

**Logs**
```
AutoSklearnClassifier(delete_output_folder_after_terminate=True,
                      delete_tmp_folder_after_terminate=False,
                      disable_evaluator_output=False,
                      ensemble_memory_limit=1024, ensemble_nbest=50,
                      ensemble_size=1, exclude_estimators=None,
                      exclude_preprocessors=None, get_smac_object_callback=None,
                      include_estimators=['random_forest'],
                      include_preprocessors=['no_preproces...
                      initial_configurations_via_metalearning=25,
                      logging_config=None, max_models_on_disc=50,
                      metadata_directory=None, ml_memory_limit=3072,
                      n_jobs=None, output_folder=None, per_run_time_limit=20,
                      resampling_strategy='holdout',
                      resampling_strategy_arguments=None, seed=1,
                      shared_mode=False, smac_scenario_args=None,
                      time_left_for_this_task=120,
                      tmp_folder='/tmp/custom_scaler')
Custom Scaler: Initialized
Custom Scaler: Initialized
Custom Scaler: Initialized
Custom Scaler: Initialized
Custom Scaler: Initialized
Custom Scaler: Initialized
accuracy:  0.958041958041958


+++++++++++++++++++++++++++++++++++++++++++Sprint+++++++++++++++++++++++++++++++++++++++++++
auto-sklearn results:
  Dataset name: breast_cancer
  Metric: accuracy
  Best validation score: 0.971631
  Number of target algorithm runs: 57
  Number of successful target algorithm runs: 56
  Number of crashed target algorithm runs: 0
  Number of target algorithms that exceeded the time limit: 1
  Number of target algorithms that exceeded the memory limit: 0



+++++++++++++++++++++++++++++++++++++++++++Show Model+++++++++++++++++++++++++++++++++++++++++++
[(1.000000, SimpleClassificationPipeline({'balancing:strategy': 'none', 'classifier:__choice__': 'random_forest', 'data_preprocessing:categorical_transformer:categorical_encoding:__choice__': 'one_hot_encoding', 'data_preprocessing:categorical_transformer:category_coalescence:__choice__': 'minority_coalescer', 'data_preprocessing:numerical_transformer:imputation:strategy': 'median', 'data_preprocessing:numerical_transformer:rescaling:__choice__': 'quantile_transformer', 'feature_preprocessor:__choice__': 'no_preprocessing', 'classifier:random_forest:bootstrap': 'True', 'classifier:random_forest:criterion': 'entropy', 'classifier:random_forest:max_depth': 'None', 'classifier:random_forest:max_features': 0.6754091314047469, 'classifier:random_forest:max_leaf_nodes': 'None', 'classifier:random_forest:min_impurity_decrease': 0.0, 'classifier:random_forest:min_samples_leaf': 3, 'classifier:random_forest:min_samples_split': 8, 'classifier:random_forest:min_weight_fraction_leaf': 0.0, 'data_preprocessing:categorical_transformer:category_coalescence:minority_coalescer:minimum_fraction': 0.02617319253950511, 'data_preprocessing:numerical_transformer:rescaling:quantile_transformer:n_quantiles': 1621, 'data_preprocessing:numerical_transformer:rescaling:quantile_transformer:output_distribution': 'uniform'},
dataset_properties={
  'task': 1,
  'sparse': False,
  'multilabel': False,
  'multiclass': False,
  'target_type': 'classification',
  'signed': False})),
]
```

I checked the cv results for the run, I can see in various runs, Custom Scaler component is assigned as params. I can attach cv results too if required.

I also tried by initialising SimpleClassificationPipeline for using custom scaler as rescaler. I got the same result. SimpleClassificationPipeline also did not used fit and transform of custom component and used default config.

## Expected behavior ##
The params assigned for the run should be used rather than using default configuration.

## Actual behavior, stacktrace or logfile ##
**Log file:** [AutoML(1)_breast_cancer.txt](https://github.com/automl/auto-sklearn/files/5017694/AutoML.1._breast_cancer.txt)

## Environment and installation: ##

Please give details about your installation:

* Is your installation in a virtual environment or conda environment? Conda environment
* Python version : 3.7.7
* Auto-sklearn version: 0.7.0

**Comment by franchuterivera** [comment]
Thanks for your question. When we fit the model, the print messages are not redirected to STDOUT.

I will recommend a better way to understand which models were tried out by auto-sklearn during fit(), which will let you also understand why a certain configuration was picked or not. Below code snapshot shows how to query autosklearn runhistory (which contains all model configurations that where tried out during fit), and in particular extracts which rescaling components where used:

```
runhistory = clf._automl[0].runhistory_
costs = []
for i, configuration in enumerate(runhistory.config_ids.keys()): 
    choices = configuration.get_dictionary()
    costs.append(
        (i, choices['data_preprocessing:numerical_transformer:rescaling:__choice__'],
        list(runhistory.data.values())[i].cost)
    )
costs.sort(key=lambda x: x[2])
```

The output applied to your code is:
```
(14, 'none', 0.021276595744680882)
(24, 'none', 0.021276595744680882)
(32, 'none', 0.021276595744680882)
(35, 'minmax', 0.021276595744680882)
(38, 'normalize', 0.021276595744680882)
(47, 'none', 0.021276595744680882)
(57, 'CustomScalerComponent', 0.021276595744680882)
...
```
In summary, we do honor the user-provided components, but the choice depends on the datasets and resources provided. With more resources in your experiments (bigger ensemble size, bigger fit time) you might be able to see this custom resizer employed in your final model.

As this doesn't seem to be a bug, I will close this issue and please re-open if further assistance is required.

**Comment by franchuterivera** [comment]
I clearly misunderstood your issue. Thanks for reporting, we are looking further into this.

**Comment by mfeurer** [comment]
Okay, @franchuterivera found the issue and we we have just merged #920 into the development branch, which should fix it and also improves testing in this direction. Please reopen this issue if the problem continue to exist..
